Enable Rack integration for Sinatra

lloeki opened the pull request:
This attemps to mimic what is currently done with Rails.

Sinatra provides only limited public facilities to manipulate the stack
(namely, appending with `use`), assumes his default middlewares are
always first, and has a peculiar way to wrap the middlewares to build
the app stack.

Sinatra also freezes the stack when the builder is run, so we have to
get access to the moment the builder is available but not yet run to
alter the list by patching `Sinatra::Base#setup_default_middlewares`.

The `Datadog::Contrib::Sinatra::Framework` class encapsulates these
internal implementation details behind a few methods to achieve the
desired manipulations.

lloeki commented:
I'm a bit unsure how to write tests for this, I tried some draft unit tests but they seemed quite artificial so I scratched them off, and testing against a Sinatra app seems tricky,  especially as Sinatra only builds the middleware stack when a request is first made. I feel like my knowledge of the various parts of the tracer test suite is a bit lacking, so I would appreciate to be either advised or helped on that matter.

lloeki commented on lib/ddtrace/contrib/sinatra/framework.rb:
```diff
@@ -0,0 +1,78 @@
+        end
+
+        def self.config_with_defaults(datadog_config)
+          datadog_config[:sinatra]
```
I picked that method from Rails and simplified it but I am not sure if it's a good way to do it.

lloeki commented on lib/ddtrace/contrib/sinatra/framework.rb:
```diff
@@ -0,0 +1,78 @@
+          datadog_config.use(
+            :rack,
+            service_name: sinatra_config[:service_name],
+            distributed_tracing: sinatra_config[:distributed_tracing],
```
Also inspired by the Rails integration. TTBOMK this seems to be the only two common configuration keys.

marcotc replied:
Given this is a fallback, it looks ok to me 👍.
The goal here is simply to make sure Rack visibility is present in tracers, but full configuration can be left to an explicit Rack configuration if the user so desires.

lloeki commented on lib/ddtrace/contrib/sinatra/patcher.rb:
```diff
@@ -1,9 +1,32 @@
+      module TracerSetupPatch
+        def setup_middleware(*args, &block)
+          super.tap do
+            Datadog::Contrib::Sinatra::Framework.setup
```
I don't know if this is the best place to call this, the Rails integration does it on `after_initialize` but I don't think there's any equivalent for Sinatra, short of patching another method. Any ideas/opinions regarding the lifecycle of the configuration?

lloeki replied:
I also saw a bunch of "only once" stuff over there for Rails. Another concern I have is whether there exists the possibility of having a Rails app that mounts Sinatra on a route, and maybe the Sinatra integration gets setup first, then the `rack.request` spans would have Sinatra as a service name instead of Rails?

I lack insight of the config/initialization lifecycle to pursue this question further. Any ideas?

marcotc replied:
> I also saw a bunch of "only once" stuff over there for Rails. Another concern I have is whether there exists the possibility of having a Rails app that mounts Sinatra on a route, and maybe the Sinatra integration gets setup first, then the rack.request spans would have Sinatra as a service name instead of Rails?

I don't think we can make a correct authoritative decision on service name in the case where two frameworks (Rails & Sinatra) trying to back fill Rack. In such case, the user will have to specify Rack's configuration explicitly. You can leave the fallback implemented in this PR as is.

> I lack insight of the config/initialization lifecycle to pursue this question further. Any ideas?

I think that we should have a `only once` guard for              `Datadog::Contrib::Sinatra::Framework.setup`. Currently it is safe to run many times (because it only conditionally configures Rack), but that might not be the case in the future.
Also, we only need to configure the fallback Rack integration once per application lifecycle.

lloeki replied:
Added OnlyOnce

lloeki commented on lib/ddtrace/contrib/sinatra/patcher.rb:
```diff
@@ -1,9 +1,32 @@
+
+      # Hook into builder before the middleware list gets frozen
+      module DefaultMiddlewarePatch
+        def setup_middleware(*args, &block)
```
`*args` for forward compatibility, in case more args are added.

marcotc commented on lib/ddtrace/contrib/sinatra/framework.rb:
```diff
@@ -0,0 +1,78 @@
+        def self.setup
+          Datadog.configure do |datadog_config|
+            sinatra_config = config_with_defaults(datadog_config)
+            activate_rack!(datadog_config, sinatra_config) unless Datadog.configuration.instrumented_integrations.key?(:rack)
+          end
```
Because `Datadog.configure` recreates all tracing internals all over again, it's good to avoid it if possible.

In this case, the conditional execution can be moved to outside the block to prevent tracer reinitialization here:

```suggestion
            activate_rack!(datadog_config, sinatra_config)
          end unless Datadog.configuration.instrumented_integrations.key?(:rack)
```

marcotc commented on lib/ddtrace/contrib/sinatra/patcher.rb:
```diff
@@ -18,12 +41,23 @@ def target_version
         end
 
         def register_tracer
           ::Sinatra.send(:register, Datadog::Contrib::Sinatra::Tracer)
           ::Sinatra::Base.prepend(Sinatra::Tracer::Base)
         end
+
+        def setup_tracer
+          ::Sinatra::Base.singleton_class.prepend(TracerSetupPatch)
+        end
+
+        def patch_default_middlewares
+          ::Sinatra::Base.singleton_class.prepend(DefaultMiddlewarePatch)
+        end
```
Any reason for not implementing the new features from `TracerSetupPatch` and `DefaultMiddlewarePatch` directly in the already existing `Sinatra::Tracer::Base`?

lloeki replied:
I wanted to semantically isolate the different patches instead of making `Sinatra::Tracer::Base` a big monolith.

lloeki commented:
Added some fixes and specs, and rebased on 0.54.2.

Do you think this should this be merged into `0.x-stable` or should it better go into `master` / `1.0-dev`?

ericmustin commented:
idk why i'm still getting pinged on this but nice work! hope y'all are well!

lloeki commented on spec/ddtrace/contrib/sinatra/tracer_spec.rb:
```diff
@@ -638,7 +631,7 @@
       it { is_expected.to be_ok }
 
-      it 'logs a warning' do
+      xit '[TODO:BROKEN:might not be possible anymore] logs a warning' do
```
I an mot sure about that one, @marcotc hypothesised it might not be possible now.

lloeki replied:
Hmm, I do see the log happening on rspec's STDOUT (even though the test case is `xit`'d):

```
Randomized with seed 3496
......**...**..................................................................................................**................................................**..........*W, [2022-01-24T11:23:36.521855 #15]  WARN -- ddtrace: [ddtrace] Sinatra integration is misconfigured, reported traces will be missing request metadata such as path and HTTP status code. Did you forget to add `register Datadog::Contrib::Sinatra::Tracer` to your `Sinatra::Base` subclass? See <https://docs.datadoghq.com/tracing/setup_overview/setup/ruby/#sinatra> for more details.
........................................**......................
```

marcotc replied:
> I an mot sure about that one, @marcotc hypothesised it might not be possible now.

Do you mean that it is not possible to have a `modular app does not register the Datadog::Contrib::Sinatra::Tracer middleware` anymore?

marcotc replied:
Found the issue, it was related to `Datadog.logger` instance being recreated between the time we setup up the test assertion vs execution time.

lloeki commented:
Rebased onto 1.0 branch, and adjusted to some namespace changes.

delner commented on lib/ddtrace/contrib/rack/middlewares.rb:
```diff
@@ -91,6 +94,9 @@ def call(env)
           raise e
         ensure
+          env[Ext::RACK_ENV_REQUEST_SPAN] = previous_request_span if previous_request_span
+          env[:datadog_rack_request_span] = env[Ext::RACK_ENV_REQUEST_SPAN]
```
I understand the above line restores the first Rack span as the current Rack span, but what does `env[:datadog_rack_request_span]` have? What's it used for?

Also, I don't think we're supposed to put symbols in the Rack env as this will break other libraries that expect all keys to be strings. Should probably convert this to a string.

delner commented on lib/ddtrace/contrib/sinatra/framework.rb:
```diff
@@ -0,0 +1,77 @@
+      module Framework
+        # Configure Rack from Sinatra, but only if Rack has not been configured manually beforehand
+        def self.setup
+          return if Datadog::Tracing.configuration.instrumented_integrations.key?(:rack)
```
Why is this guard necessary? When does it run multiple times? I would think if it ran configuration twice this shouldn't be an issue. I can see this being especially problematic for tests where it generates and re-applies patches between examples and monkey-patches hang around (in which case it won't run the configure step.)

lloeki replied:
I was presuming that if Rack was configured manually by the user we might want to keep that configuration instead of reconfiguring it, possibly overrwriting/resetting the user provided configuration.

lloeki replied:
Removed for now, for consistency with the Rails integration.

delner commented on lib/ddtrace/contrib/sinatra/framework.rb:
```diff
@@ -0,0 +1,77 @@
+        def self.activate_rack!(datadog_config, sinatra_config)
+          datadog_config.instrument(
+            :rack,
+            service_name: sinatra_config[:service_name],
```
Should Rack receive Sinatra's service name? Generally speaking, I don't think this is an issue when in most cases we want all instrumentation to reflect the same service name. However, this will not allow users to override Rack service names unless they do it after this configuration occurs.

marcotc replied:
This is consistent with the Rails instrumentation today: https://github.com/DataDog/dd-trace-rb/blob/93cd757770df59455fc94d04bb2db62f3dd22ada/lib/ddtrace/contrib/rails/framework.rb#L73-L76

It seems harsh to not allow `c.use :rack, service_name:` to take place, but it's what Rails does.

There's an argument that allowing two different service names between Rack and Rails/Sinatra is asking for confusion.

TBH, I believe we are moving towards removing the ability to configure custom service names for internal integrations separately, so having them have the same service name is, in a roundabout way, aligned with our simplified service naming goals.

lloeki replied:
Indeed as a general trend in this PR I tried to port over how the Rails integration behaves, taking it as a baseline for behaviour.

It is arguable whether Rack should be considered an independent service or part of the service that depends on it (Sinatra, Rails), but I would argue for the latter, not the least because the frameworks pick specific Rack middlewares from the `rack` gem and interleave these with their own, so the framework implements itself not just "over"/"after" Rack but "with" Rack.

lloeki replied:
> I believe we are moving towards removing the ability to configure custom service names for internal integrations separately

That was the impetus for the guard glause, which according to what you say, might be irrelevant:

```
return if Datadog::Tracing.configuration.instrumented_integrations.key?(:rack)
```

Come to think of it, in the case of nested apps, I wonder how configuring the Rack integration service name would be expected to work, if at all (I didn't check but intuitively it seems all Rack services would be named the same, nested or not?)

delner commented on lib/ddtrace/contrib/sinatra/patcher.rb:
```diff
@@ -1,9 +1,40 @@
     module Sinatra
+      # Set tracer configuration at a late enough time
+      module TracerSetupPatch
+        ONLY_ONCE_PER_APP = Hash.new { |h, key| h[key] = Datadog::Utils::OnlyOnce.new }
```
In most scenarios this is probably innocuous, but part of me is concerned about a `Hash` that accumulates `app` references as keys and holds onto them (presumably) indefinitely. Is there any concern for a memory leak?

Is there a better way of doing this? Make the "only_once" an instance variable on the instance perhaps?

marcotc replied:
I think this number is strictly limited by number of Sinatra applications exist. This number will be very small, the vast majority of time 1. I'd be surprised to see this number > 10 across all our users.

lloeki replied:
I concur with @marcotc here. I took this pattern from the Rails integration: see `BEFORE_INITIALIZE_ONLY_ONCE_PER_APP`, `AFTER_INITIALIZE_ONLY_ONCE_PER_APP`.

delner commented on lib/ddtrace/contrib/sinatra/tracer_middleware.rb:
```diff
@@ -55,6 +55,16 @@ def call(env)
               span.resource = env['REQUEST_METHOD'] if span.resource.nil?
 
+              rack_request_span = env[Datadog::Contrib::Rack::Ext::RACK_ENV_REQUEST_SPAN]
+              if rack_request_span
```
Minor, but is there a more concise way of expressing this logic? `rack_request_span.resource ||= span.resource if rack_request_span`?

lloeki replied:
Indeed. This is a carryover from a conflict, there used to be an `if`/`else` clause instead of unless in this `if`.

delner reviewed (commented):
Just some questions for now!

marcotc commented on lib/ddtrace/contrib/rack/middlewares.rb:
```diff
@@ -49,16 +49,19 @@ def compute_queue_time(env)
 
         def call(env)
+          # Find out if this is rack within rack
+          previous_request_span = env[Ext::RACK_ENV_REQUEST_SPAN]
```
Does Rack actually get called recursively? It's very 👻 if so, because the Rack instrumentation does stuff like this:
https://github.com/DataDog/dd-trace-rb/blob/93cd757770df59455fc94d04bb2db62f3dd22ada/lib/ddtrace/contrib/rack/middlewares.rb#L123

lloeki replied:
Nesting whole rack (or sinatra) apps within rack (or sinatra, or rails) apps is definitely possible, and actually tested by the specs. Without this kind of fetch/check/restore (at `call`'s `ensure`), consequences appeared in the spec, e.g distributed tracing headers being mistakenly picked up by the deeper rack integration middleware, severing them from their actual parent.

The mentioned reset line I catered for when based on 0.x, but it appears it does not exist anymore on 1.0, the corresponding setter at `call` beginning became `Datadog::Tracing.continue_trace!(trace_digest)`.

I surmised the following comment above the line you quoted was no longer relevant as this may now better handled by the new trace concepts:

```
          # TODO: Remove this once we change how context propagation works. This
          # ensures we clean thread-local variables on each HTTP request avoiding
          # memory leaks.
```

But do tell me if I am mistaken as these concepts are very new to me.

lloeki commented:
Rebased on latest 1.0 to get the msgpack fix on JRuby, but I got a spec failure... and this failure is also present when I checkout origin/1.0 (2eae95e81) when it was not on 6154c7ea1 (my previous base):

```

Randomized with seed 51120
........................................................................F..................

Failures:

  1) Rack integration tests for an application with a route that leaks span context subsequent GET requests is expected to equal nil
     Failure/Error: expect(tracer.active_trace).to be nil
     
       expected nil
            got #<Datadog::TraceOperation:69860> => #<Datadog::TraceOperation:0x00005633adc432e0 @events=#<Datadog::TraceOperation::Events:0x00005633adc4...rg", "http.status_code"=>"200"}>, @active_span=nil, @active_span_count=0, @finished=true, @spans=[]>
     # ./spec/ddtrace/contrib/rack/integration_test_spec.rb:522:in `block (5 levels) in <top (required)>'
     # ./spec/ddtrace/contrib/support/tracer_helpers.rb:94:in `block (2 levels) in <module:TracerHelpers>'
     # ./spec/spec_helper.rb:207:in `block (2 levels) in <top (required)>'
     # ./spec/spec_helper.rb:99:in `block (2 levels) in <top (required)>'
     # /usr/local/bundle/gems/webmock-3.13.0/lib/webmock/rspec.rb:37:in `block (2 levels) in <top (required)>'

Finished in 0.98924 seconds (files took 1.96 seconds to load)
91 examples, 1 failure

Failed examples:

rspec ./spec/ddtrace/contrib/rack/integration_test_spec.rb:520 # Rack integration tests for an application with a route that leaks span context subsequent GET requests is expected to equal nil
```

marcotc commented on spec/ddtrace/contrib/sinatra/multi_app_spec.rb:
```diff
@@ -69,8 +69,15 @@
         it do
           is_expected.to be_ok
-          expect(spans).to have(2).items
+          expect(spans).to have(3).items
```
Is the extra span the Rack span?
Does this mean that users upgrading to the code in this PR will see an extra span in their trace?

lloeki replied:
Yes, and yes.
